### PR TITLE
Clean repository ROOT directory name with filepath.Clean

### LINKED ISSUE #2845: Windows UNC paths do not work for some config settings
- Gitea version (or commit ref): 1.2.3
- Git version: 2.15.0
- Operating system: Windows Server 2012 & Windows 7
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log message:

Something like:

2017/11/03 13:22:19 [...ules/context/repo.go:335 func1()] [E] RepoAssignment Invalid repo \mycifsserver\mywindowsshare\repositories\orgname\reponame.git: no such file or directory

## Description

It is impossible to use windows UNC names (aka windows file shares) in the repository ROOT setting.  The UNC name has the double slashes at the start converted to a single slash.  For example:

```
[repository]
ROOT = //mycifsserver/mywindowsshare/repositories
```

Will be converted into:

```
\mycifsserver\mywindowsshare\repositories
```

I've inspected the code and the `path` package is used in many places to Clean or Join paths, including where the ROOT setting is read.  `path` will not do the right thing on Windows, instead `filepath` should be used.  From the docs for path:

<em>
Package path implements utility routines for manipulating slash-separated paths.
<br><br>
The path package should only be used for paths separated by forward slashes, such as the paths in URLs. This package does not deal with Windows paths with drive letters or backslashes; to manipulate operating system paths, use the path/filepath package.
</em>

### PR DESCRIPTION
**Pull request by MCF**
For better compatibility with Windows.

Allows UNC paths for the repository ROOT.  Fixes: https://github.com/go-gitea/gitea/issues/2845

### CODE REVIEW

**Comment by MCF** [comment]
FWIW I'm already running this patch in a windows production environment.  Although I applied it to the release/V1.2 branch in that case.  I'm happy to provide a backport to that branch if/when this PR is accepted.

File: modules/setting/setting.go [hunk 1]
```diff
@@ -973,7 +973,7 @@ func NewContext() {
 	if !filepath.IsAbs(RepoRootPath) {
 		RepoRootPath = path.Join(AppWorkPath, RepoRootPath)
```
Comment (lafriks) [inline modules/setting/setting.go]: Maybe change also here to `filepath.Join` so that behaviour is the same in both cases

**Comment by MCF** [comment]
@lafriks I've made the change you requested.

I think it is important to clarify that only filepath.Clean is really important in the context of paths based on settings that come out of app.ini.  As mentioned in the related issue #2845, app.ini only allows front slashes in the repository ROOT setting.  This is enforced by `forcePathSeparator`, which is used for a number of other settings:

```bash
$ grep forcePathSeparator modules/setting/setting.go 
func forcePathSeparator(path string) {
	forcePathSeparator(LogRootPath)
	forcePathSeparator(RepoRootPath)
	forcePathSeparator(AvatarUploadPath)
```
As a result the calls to path.Join vs. filepath.Join don't matter so much in this context.  For example:

```
[repository]
ROOT = store/repos    ; front slash required.
```

On Windows with the a GITEA work dir of `C:\gitea` the resulting paths would be:

filepath.Join `C:\gitea\store/repos`  
path.Join `C:\gitea/store/repos`

Fortunately either variant works with go's file and directory functions.  Go seems very forgiving when it comes to this sort of problem and seems to accept either slash in path names on Windows.

TL;DR I don't think we have to worry about path.Join, only path.Clean when it comes to UNC file system paths (or any other file paths) on Windows.

**Comment by lafriks** [comment]
LGTM

**Comment by Morlinest** [comment]
LGTM
